x/stakeibc/keeper/icacallbacks.go

```diff
@@ -116,3 +120,60 @@ func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
 	k.RecordsKeeper.RemoveDepositRecord(ctx, cast.ToUint64(recordId))
 	return nil
 }
+
+// ----------------------------------- Reinvest Callback ----------------------------------- //
+func (k Keeper) MarshalReinvestCallbackArgs(ctx sdk.Context, delegateCallback types.ReinvestCallback) ([]byte, error) {
+	out, err := proto.Marshal(&delegateCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("MarshalReinvestCallbackArgs %v", err.Error()))
+		return nil, err
+	}
+	return out, nil
+}
+
+func (k Keeper) UnmarshalReinvestCallbackArgs(ctx sdk.Context, delegateCallback []byte) (*types.ReinvestCallback, error) {
+	unmarshalledReinvestCallback := types.ReinvestCallback{}
+	if err := proto.Unmarshal(delegateCallback, &unmarshalledReinvestCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalReinvestCallbackArgs %v", err.Error()))
+		return nil, err
+	}
+	return &unmarshalledReinvestCallback, nil
+}
+
+func ReinvestCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack *channeltypes.Acknowledgement_Result, args []byte) error {
+	// invariant: sendMsg.FromAddress == withdrawalAddress && sendMsg.ToAddress == delegationAddress
```

Comment:
is this something we should confirm here?

Reply:
+1: the [tx you submit with this callback](https://github.com/Stride-Labs/stride/pull/182/files?diff=split&w=0#diff-b6597a6df3c7e86ae53cb25938eab9ea9e920ea686410592068c6deaf4cae116R163) contains two messages : one to the revenue acct and one to the delegation acct. where do we distinguish between one and the other in the CB?

Reply:
The callback executes at the transaction level, so you don't need to distinguish between messages (both succeed or fail atomically as part of the transaction)

We have so much sensitive logic in callbacks that I'm not sure how much it would add to check the invariant